core/src/main/scala/spinal/core/Enum.scala

```diff
@@ -349,6 +349,29 @@ object binaryOneHot extends SpinalEnumEncoding{
   setName("oh")
 }
 
+/**
+ * Gray encoding (sequentially assigned)
+ * @example{{{ 000, 001, 011, 010, ... }}}
+ * @note If used in FSM is is not ensured that only gray encoding preserving
+ *       transitions are done. If that is needed e.g. for CDC reasons, the
+ *       transitions must be checked manually.
+ */
+object graySequential extends SpinalEnumEncoding {
+  private def gray2binary(gray: BigInt) = {
+    assert(gray.bitLength <= 16, "conversion currently only implemented for <= 16 bit")
+    var tmp = gray ^ (gray >> 8)
+    tmp ^= tmp >> 4
+    tmp ^= tmp >> 2
+    tmp ^= tmp >> 1
+    tmp
+  }
+  private def binary2gray(bin: BigInt) = bin ^ (bin >> 1)
+
+  override def getWidth(e: SpinalEnum) = log2Up(e.elements.length)
+  override def getValue[T <: SpinalEnum](element: SpinalEnumElement[T]): BigInt = binary2gray(element.position)
+  override def getElement[T <: SpinalEnum](value: BigInt, enums: T): SpinalEnumElement[T] = enums.elements(binary2gray(value).toInt)
```

Comment:
I'm wondering, shouldn't it be `gray2binary` here?

Reply:
There is a "toGray" and "fromGray" [here](https://github.com/SpinalHDL/SpinalHDL/blob/dda945dfafa3e479c680a96385a94bd8e54f929a/lib/src/main/scala/spinal/lib/Utils.scala#L369).
Although, it's a spinal one, but not the Scala function, it's still a good reference.